File: Leger/PolyPaint/app/src/main/java/com/polypaint/polypaint/Model/Link.kt
package com.polypaint.polypaint.Model

import java.io.Serializable
import java.util.*
import kotlin.collections.ArrayList

class Link(var id: String, var name: String, var from: AnchorPoint, var to: AnchorPoint, var type: Int, var style: LinkStyle, var path: ArrayList<Coordinates>) : Serializable, DrawingElement(id){


    fun copy(id: String = UUID.randomUUID().toString(), name: String= this.name, from: AnchorPoint = AnchorPoint(), to: AnchorPoint = AnchorPoint(), type:Int = this.type, style: LinkStyle=this.style.copy(), path: ArrayList<Coordinates> = copyPathCoordinates()) = Link(id,name, from, to,type,style,path)

    private fun copyPathCoordinates():ArrayList<Coordinates>{
        var pathCopied : ArrayList<Coordinates> = ArrayList()
        for(coordinate in this.path){
            pathCopied.add(coordinate.copy())
        }
        return pathCopied
    }
}

File: Lourd/PolyPaint/PolyPaint/Templates/Link.cs
using PolyPaint.CustomInk;
using System;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Ink;

namespace PolyPaint.Templates
{
    public class Link
    {
        public string id { get; set; }
        public string name { get; set; }
        public AnchorPoint from { get; set; }
        public AnchorPoint to { get; set; }
        public int type { get; set; }
        public LinkStyle style { get; set; }
        public List<Coordinates> path { get; set; }

        public Link()
        {
        }

        public Link(string id, string name, AnchorPoint from, AnchorPoint to, int type, LinkStyle style, List<Coordinates> path)
        {
            this.id = id;
            this.name = name;
            this.from = from;
            this.to = to;
            this.type = type;
            this.style = style;
            this.path = path;
        }
        
        public Point GetPoint(int indexInPath)
        {
            return new Point(path[indexInPath].x, path[indexInPath].y);
        }

        public Point GetFromPoint(StrokeCollection strokes)
        {
            CustomStroke fromStroke;
            Point point = new Point();

            foreach(CustomStroke stroke in strokes)
            {
                if (stroke.guid.ToString() == this.from.formId)
                {
                    fromStroke = stroke;
                    point = (fromStroke as ShapeStroke).GetAnchorPoint(this.from.anchor);
                    //point = (fromStroke as ShapeStroke).anchorPoints[from.anchor];
                }
            }

            return point;
        }

        public Point GetToPoint(StrokeCollection strokes)
        {
            CustomStroke fromStroke;
            Point point = new Point();

            foreach (CustomStroke stroke in strokes)
            {
                if (stroke.guid.ToString() == this.to.formId)
                {
                    fromStroke = stroke;
                    point = (fromStroke as ShapeStroke).GetAnchorPoint(this.to.anchor);
                }
            }

            return point;
        }
    }
}